File: solutions/clojure/cars-assemble/1/src/cars_assemble.clj
(ns cars-assemble)

(def gross-production-per-hour 221)

(defn success-rate
  [speed]
  (cond (= speed 0) 0.0
        (<= speed 4) 1.0
        (<= speed 8) 0.90
        (<= speed 9) 0.8
        :else 0.77))

(* gross-production-per-hour 5)

(defn production-rate
  "Returns the assembly line's production rate per hour,
   taking into account its success rate"
  [speed]
  (* (* gross-production-per-hour
        speed)
     (success-rate speed)))

(production-rate 6)

(defn working-items
  "Calculates how many working cars are produced per minute"
  [speed]
  (int 
   (/ (production-rate speed) 60)
   )
  )

(working-items 6)

File: solutions/go/cars-assemble/1/cars_assemble.go
package cars

// CalculateWorkingCarsPerHour calculates how many working cars are
// produced by the assembly line every hour.
func CalculateWorkingCarsPerHour(productionRate int, successRate float64) float64 {
	return float64(productionRate) * successRate / 100.0
}

// CalculateWorkingCarsPerMinute calculates how many working cars are
// produced by the assembly line every minute.
func CalculateWorkingCarsPerMinute(productionRate int, successRate float64) int {
	carsPerMinute := CalculateWorkingCarsPerHour(productionRate, successRate) / 60.0
	return int(carsPerMinute)
}

// CalculateCost works out the cost of producing the given number of cars.
func CalculateCost(carsCount int) uint {
	const costPer10 = 95000
	const costPer1 = 10000
	return uint(carsCount / 10 * costPer10 + carsCount % 10 * costPer1)
}
